hasher-matcher-actioner/hmalib/fetching/tests/test_fetcher_and_bank_store.py

```diff
@@ -0,0 +1,310 @@
+
+    table = dynamodb.create_table(**BanksTableTestBase.get_table_definition())
+    yield BanksTable(table, get_default_signal_type_mapping())
+    mddb.stop()
+
+
+@pytest.fixture
+def mock_enabled_signal_exchange_apis(monkeypatch):
+    def mock_get(*args, **kwargs):
+        return [
+            ToggleableSignalExchangeAPIConfig(
+                "fake",
+                "hmalib.fetching.tests.test_fetcher_and_bank_store.FakeAPI",
+                True,
+            )
+        ]
+
+    monkeypatch.setattr(ToggleableSignalExchangeAPIConfig, "get_all", mock_get)
+
+
+def get_fake_collab_config(import_as_bank_id: str) -> CollaborationConfigBase:
+    pytx_config = CollaborationConfigWithDefaults("Test State", "fake")  # type: ignore
+
+    return EditableCollaborationConfig(
+        name="does not matter",
+        collab_config_class="threatexchange.exchanges.collab_config.CollaborationConfigWithDefaults",
+        attributes_json_serialized=dataclass_dumps(pytx_config),
+        import_as_bank_id=import_as_bank_id,
+    )
+
+
+# For brevity when describing test cases.
+K1 = "key-one"
+K2 = "key-two"
+K3 = "key-three"
+
+
+def md5(n: int) -> str:
+    return f"{n:032x}"
+
+
+def pdq(n: int) -> str:
+    return f"{n:064x}"
+
+
+# Each test case is a 3 value tuple. For each case n,
+# tuple[0] is a human readable test name. We'll print this if tests fail.
+# tuple[1] is the update to be emitted by fetch_iter()
+# tuple[2] is the expected end state after all n tuple[1]s have been emitted.
+# tuple[3] if NotImplemented, indicates unsupported behavior.
+#
+# Note: the test cases are cumulative. Each test case will behave as if all the
+# previous tests have been performed on this store.
+TEST_CASES = [
+    (
+        "Must have no indexable content when naive_convert_to_signal_type returns None",
+        {K1: FakeUpdateRecord(ImaginarySignalType, md5(1), "tag-one")},
+        {},
+    ),
+    (
+        "Must have one indexable signal for one known signal type.",
+        {K1: FakeUpdateRecord(VideoMD5Signal, md5(1), "tag-one")},
+        {VideoMD5Signal: {md5(1)}},
+    ),
+    (
+        "Must be able to replace signals of same key..",
+        {
+            K1: FakeUpdateRecord(VideoMD5Signal, md5(2), "tag-two"),
+            K2: FakeUpdateRecord(VideoMD5Signal, md5(3), "tag-three"),
+        },
+        {VideoMD5Signal: {md5(3)}},
+        NotImplemented,
+    ),
+    (
+        "Must be able to handle a delete for a known key.",
+        {K1: None},
+        {VideoMD5Signal: {md5(3)}},
+    ),
+]
+
+
+def unwrap_bank_signals_response(response: PaginatedResponse) -> t.Set[str]:
+    # Helper to convert bank member signals response to something that can be
+    # compared with TEST_CASE data
+    return {item.signal_value for item in response.items}
+
+
+@pytest.mark.usefixtures(
+    "mock_enabled_signal_exchange_apis",
+)
+@pytest.mark.parametrize("i", range(len(TEST_CASES)))
+def test_for_case_until(i, banks_table):
```

Comment:
I swallowed the pill!

functional tests are, in fact, much more natural and conducive to fixtures, and parameters and all the good things.

Reply:
For me, having the assert functions provide more context was the thing that motived me to start rewriting the old ones.

Otherwise, anything you can do with functional tests you can do with test case if you are motivated enough, but mix and matching is very nice.